pkg/oci/builder_test.go

```diff
@@ -172,8 +179,82 @@ func validateOCI(path string, t *testing.T) {
 		t.Fatalf("invalid schema version, expected 2, got %d", imageIndex.SchemaVersion)
 	}
 
-	// Additional validation of the Image Index structure can be added here
-	// extract. for example checking that the path includes the README.md
-	// and one of the binaries in the exact location expected (the data layer
-	// blob and exec layer blob, respectively)
+	if len(imageIndex.Manifests) < 1 {
+		t.Fatal("fewer manifests")
+	}
+
+	digest := strings.TrimPrefix(imageIndex.Manifests[0].Digest, "sha256:")
+	manifestFile := filepath.Join(path, "blobs", "sha256", digest)
+	manifestFileData, err := os.ReadFile(manifestFile)
+	if err != nil {
+		t.Fatal(err)
+	}
+	mf := struct {
+		Layers []struct {
+			Digest string `json:"digest"`
+		} `json:"layers"`
+	}{}
+	err = json.Unmarshal(manifestFileData, &mf)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	type fileInfo struct {
+		Path       string
+		Type       fs.FileMode
+		Executable bool
+	}
+	var files []fileInfo
+
+	for _, layer := range mf.Layers {
+		func() {
+			digest = strings.TrimPrefix(layer.Digest, "sha256:")
+			f, err := os.Open(filepath.Join(path, "blobs", "sha256", digest))
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer f.Close()
+
+			gr, err := gzip.NewReader(f)
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer gr.Close()
+
+			tr := tar.NewReader(gr)
+			for {
+				hdr, err := tr.Next()
+				if err != nil {
+					if errors.Is(err, io.EOF) {
+						break
+					}
+					t.Fatal(err)
+				}
+				files = append(files, fileInfo{
+					Path:       hdr.Name,
+					Type:       hdr.FileInfo().Mode() & fs.ModeType,
+					Executable: (hdr.FileInfo().Mode()&0111 == 0111) && !hdr.FileInfo().IsDir(),
+				})
+			}
+		}()
+	}
+	sort.Slice(files, func(i, j int) bool {
+		return files[i].Path < files[j].Path
+	})
+
+	expectedFiles := []fileInfo{
+		{Path: "/etc/pki/tls/certs/ca-certificates.crt"},
+		{Path: "/etc/ssl/certs/ca-certificates.crt"},
+		{Path: "/func", Type: fs.ModeDir},
+		{Path: "/func/README.md"},
+		{Path: "/func/f", Executable: true},
+		{Path: "/func/func.yaml"},
+		{Path: "/func/go.mod"},
+		{Path: "/func/handle.go"},
+		{Path: "/func/handle_test.go"},
+	}
```

Comment:
@lkingland why are the sources in the image? Shouldn't be there only executable and certificates?

Reply:
The third layer is the `data` layer (for example, a repository which is a static HTTP site and a .go function which serves the files on request).  The goal here was to use a `.funcignore` pre-populated with `*.go` when that feature is available.  I will open a separate issue for that

Reply:
@lkingland I am afraid that if you add `*.go`s to `.funcignore` then the source will be also ignored by builders -- it won't compile application at all.

Reply:
It could, depending on how we interpret the .funcignore per builder

I [opened a discussion](https://github.com/knative/func/discussions/1876)